src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -133,9 +134,10 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
                         ProcessRegistrationPage(registrationPage, results, range, includePrerelease, includeUnlisted, metadataCache);
                     }
                 }
-            }
+            });
 
-            return results;
+            await Task.WhenAll(tasks);
+            return results.OrderBy(result => result.Version.Version).ToList();
```

Comment:
The general concern so far is that changing the order of results is a behavioral change we want to avoid if it is feasible.

The order is defined by the server, so even an explicit ordering like this can change the server's ordering.  Also, I think it's possible to have different orderings if you sort by NuGetVersion versus NuGetVersion.Version (System.Version) because System.Version doesn't contain build or metadata information.

How about something [like this](https://github.com/NuGet/NuGet.Client/commit/5d556957dbec122a65b34dd51326042c02a46aed) to preserve server ordering?  (Disclaimer:  I didn't actually test this code.  I only mentally executed it.)

In terms of performance, I don't know how this compares with other options, so performance testing is necessary.